File: brainfuck/bf.cc
#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>
#include <stack>
#include <map>
#include <fstream>

#include <unistd.h>

using namespace std;

class Program
{
    string code;
    vector<size_t> bracket_pc;

public:
    Program(const string &source)
    {
        this->code.reserve(source.size());
        this->bracket_pc.resize(source.size(), 0);

        stack<size_t> stack;

        for (char ch : source)
        {
            if (ch == '[')
            {
                stack.push(this->code.size());
            }
            else if (ch == ']')
            {
                size_t left = stack.top();
                stack.pop();
                size_t right = this->code.size();
                this->bracket_pc[left] = right;
                this->bracket_pc[right] = left;
            }
            else
            {
                if (ch != '<' && ch != '>' && ch != '+' && ch != '-' && ch != '.')
                {
                    continue;
                }
            }

            this->code.push_back(ch);
        }
    }

    void run()
    {
        vector<unsigned char> tape{0};
        tape.reserve(8196);

        size_t pc = 0;
        size_t ptr = 0;

        size_t end = this->code.size();

        while (pc < end)
        {
            char ch = this->code[pc];
            if (ch == '+')
            {
                tape[ptr] += 1;
            }
            else if (ch == '-')
            {
                tape[ptr] -= 1;
            }
            else if (ch == '>')
            {
                ptr += 1;
                if (tape.size() == ptr)
                {
                    tape.push_back(0);
                }
            }
            else if (ch == '<')
            {
                ptr -= 1;
            }
            else if (ch == '.')
            {
                putchar(tape[ptr]);
                // printf("%c", tape[ptr]);
            }
            else if (ch == '[' && tape[ptr] == 0)
            {
                pc = this->bracket_pc[pc];
            }
            else if (ch == ']' && tape[ptr] != 0)
            {
                pc = this->bracket_pc[pc];
            }
            pc += 1;
        }
    }
};

int main(int argc, char *argv[])
{
    std::ifstream source_file(getenv("GM_BF_FILE"));

    if (!source_file.is_open()) {
        return -1;
    }

    source_file.seekg(0, std::ios::end);

    string source;

    size_t size = source_file.tellg();
    source.resize(size);
    source_file.seekg(0);
    source_file.read(&source[0], size);

    Program program(source);

    program.run();

    return 0;
}

File: brainfuck/bf.rs
use std::env;
use std::io::{self, BufWriter, Write};
use std::iter;

struct Program<W: Write> {
    source: Vec<u8>,
    bracket_pc: Vec<usize>,
    out: W,
}

impl<W: Write> Program<W> {
    pub fn new(source: &[u8], out: W) -> Self {
        let mut bf_source = Vec::with_capacity(source.len());
        let mut bracket_pc: Vec<usize> = iter::repeat(0).take(source.len()).collect();
        let mut stack = Vec::with_capacity(10);

        for s in source.iter() {
            match s {
                b'>' | b'<' | b'+' | b'-' | b',' | b'.' => {}
                b'[' => {
                    stack.push(bf_source.len());
                    bf_source.push(*s);
                }
                b']' => {
                    let left = stack.pop().unwrap();
                    let right = bf_source.len();
                    bf_source.push(*s);
                    bracket_pc[left] = right;
                    bracket_pc[right] = left;
                }
                _ => continue,
            }
            bf_source.push(*s);
        }

        Self {
            out,
            source: bf_source,
            bracket_pc,
        }
    }

    pub fn run(&mut self) {
        let mut pc = 0;
        let mut ptr = 0;
        let mut tape = vec![0; 8096];
        while let Some(byte) = self.source.get(pc).copied() {
            match byte {
                b'>' => {
                    ptr += 1;
                    if ptr >= tape.len() {
                        tape.push(0);
                    }
                }
                b'<' => {
                    ptr -= 1;
                }
                b'+' => {
                    tape[ptr] += 1;
                }
                b'-' => {
                    tape[ptr] -= 1;
                }
                b'.' => {
                    self.out.write(&[tape[ptr]]).unwrap();
                }
                b'[' => {
                    if tape[ptr] == 0 {
                        pc = self.bracket_pc[pc];
                    }
                }
                b']' => {
                    if tape[ptr] != 0 {
                        pc = self.bracket_pc[pc];
                    }
                }
                _ => {}
            }

            pc += 1;
        }
    }
}

fn main() {
    let stdout = io::stdout();
    let stdout = stdout.lock();
    let mut stdout = BufWriter::with_capacity(8196, stdout);

    let source_file = env::var("GM_BF_FILE").unwrap();
    let source = std::fs::read(source_file).unwrap();

    let mut program = Program::new(&source, &mut stdout);

    program.run();
}
